### ISSUE TITLE
deprecate POSIX::tmpnam

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13892]
Migrated from [rt.perl.org#122005](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=122005) (status was 'resolved')

Searchable as RT122005$

**Comment by p5pRT** [comment]
### From @jhi

#### Created by @jhi



The tmpnam\(\) interface is known to be dangerous enough that compilers 
and even linkers warn against its use\.  And POSIX\.pod advises likewise\.

Therefore\, POSIX extension probably should stop giving access to the raw 
tmpnam\(\)\.

Some options&#8203;:

\(1\) remove it\, let users \(\*\) die
\(2\) implement it using mkstemp\(\) if available\.  \(What about platforms 
that have tmpnam\(\) but do not have mkstemp\(\)?\)  This reimplementing\, of 
course\, is already done \(and much more\) in File&#8203;::Temp\, but we probably 
don't want the POSIX being dependent on File&#8203;::Temp\.

Both options above can include "add a loud obnoxious warning"\.

\(\*\) http&#8203;://grep\.cpan\.me/?q=POSIX\.\*tmpnam

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.21.1:

Configured by jhi at Sat May 31 08:28:57 EDT 2014.

Summary of my perl5 (revision 5 version 21 subversion 1) configuration:
  Commit id: c56ffef01cadd0c70cc6efca30bf8b6d5ba2e312
  Platform:
    osname=darwin, osvers=13.2.0, archname=darwin-2level
    uname='darwin vredefort-2.local 13.2.0 darwin kernel version 13.2.0: thu apr 17 23:03:13 pdt 2014; root:xnu-2422.100.13~1release_x86_64 x86_64 '
    config_args='-des -Dusedevel'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -I/opt/local/include',
    optimize='-O3',
    cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -I/opt/local/include'
    ccversion='', gccversion='4.2.1 Compatible Apple LLVM 5.1 (clang-503.0.40)', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' -fstack-protector -L/usr/local/lib -L/opt/local/lib'
    libpth=/usr/local/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/5.1/lib /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib /usr/lib /opt/local/lib
    libs=-lgdbm -ldbm -ldl -lm -lutil -lc
    perllibs=-ldl -lm -lutil -lc
    libc=, so=dylib, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup -L/usr/local/lib -L/opt/local/lib -fstack-protector'



@INC for perl 5.21.1:
    lib
    /usr/local/lib/perl5/site_perl/5.21.1/darwin-2level
    /usr/local/lib/perl5/site_perl/5.21.1
    /usr/local/lib/perl5/5.21.1/darwin-2level
    /usr/local/lib/perl5/5.21.1
    .


Environment for perl 5.21.1:
    DYLD_LIBRARY_PATH (unset)
    HOME=/Users/jhi
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/Users/jhi/bin:/opt/local/perl/bin:/usr/local/git/bin:/opt/bin:/opt/local/bin:/Users/jhi/.s:/opt/bin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin:/usr/libexec:/usr/lib:/etc:/Users/jhi:/Applications/MacPorts/Eiffel60/studio/spec/macosx-x86/bin:/Applications/MacPorts/Eiffel60/library/gobo/spec/macosx-x86/bin
    PERL_BADLANG (unset)
    SHELL=/opt/local/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Sun Jun 01 06&#8203;:27&#8203;:39 2014\, jhi wrote&#8203;:
> The tmpnam\(\) interface is known to be dangerous enough that compilers
> and even linkers warn against its use\.  And POSIX\.pod advises
> likewise\.
> 
> Therefore\, POSIX extension probably should stop giving access to the
> raw
> tmpnam\(\)\.
> 
> Some options&#8203;:
> 
> \(1\) remove it\, let users \(\*\) die
>  \(2\) implement it using mkstemp\(\) if available\.  \(What about platforms
>  that have tmpnam\(\) but do not have mkstemp\(\)?\)  This reimplementing\,
> of
>  course\, is already done \(and much more\) in File&#8203;::Temp\, but we
> probably
> don't want the POSIX being dependent on File&#8203;::Temp\.
> 
> Both options above can include "add a loud obnoxious warning"\.
> 
> \(\*\) http&#8203;://grep\.cpan\.me/?q=POSIX\.\*tmpnam

\(3\) produce a deprecation warning\, patch attached\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-deprecate-POSIX-tmpnam.patch</summary>

```diff
From a35462eee8412e905fcbeaad3ee92d794ea50ba0 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Tue, 22 Jul 2014 11:44:54 +1000
Subject: [PATCH] deprecate POSIX::tmpnam()

---
 ext/POSIX/POSIX.xs     |    2 ++
 ext/POSIX/lib/POSIX.pm |    2 +-
 ext/POSIX/t/posix.t    |   14 +++++++++++++-
 3 files changed, 16 insertions(+), 2 deletions(-)

diff --git a/ext/POSIX/POSIX.xs b/ext/POSIX/POSIX.xs
index 2a77df0..e3a6f83 100644
--- a/ext/POSIX/POSIX.xs
+++ b/ext/POSIX/POSIX.xs
@@ -1505,6 +1505,8 @@ tmpnam()
 	STRLEN i;
 	int len;
     CODE:
+        if (ckWARN_d(WARN_DEPRECATED))
+	    Perl_warner(aTHX_ packWARN(WARN_DEPRECATED), "Calling POSIX::tmpnam() is deprecated");
 	RETVAL = newSVpvs("");
 	SvGROW(RETVAL, L_tmpnam);
 	/* Yes, we know tmpnam() is bad.  So bad that some compilers
diff --git a/ext/POSIX/lib/POSIX.pm b/ext/POSIX/lib/POSIX.pm
index 57845a7..3daa2f3 100644
--- a/ext/POSIX/lib/POSIX.pm
+++ b/ext/POSIX/lib/POSIX.pm
@@ -4,7 +4,7 @@ use warnings;
 
 our ($AUTOLOAD, %SIGRT);
 
-our $VERSION = '1.41';
+our $VERSION = '1.42';
 
 require XSLoader;
 
diff --git a/ext/POSIX/t/posix.t b/ext/POSIX/t/posix.t
index c2e4abe..3d68d76 100644
--- a/ext/POSIX/t/posix.t
+++ b/ext/POSIX/t/posix.t
@@ -8,7 +8,7 @@ BEGIN {
     }
 }
 
-use Test::More tests => 109;
+use Test::More tests => 111;
 
 use POSIX qw(fcntl_h signal_h limits_h _exit getcwd open read strftime write
 	     errno localeconv dup dup2 lseek access);
@@ -401,6 +401,18 @@ SKIP: {
     cmp_ok($!, '==', POSIX::ENOTDIR);
 }
 
+{   # tmpnam() is deprecated
+    my $warn;
+    local $SIG{__WARN__} = sub { $warn .= "@_" };
+    my $foo = POSIX::tmpnam();
+    like($warn, qr!^Calling POSIX::tmpnam\(\) is deprecated at t/posix.t line \d+\.$!,
+       "check POSIX::tmpnam warns by default");
+    no warnings "deprecated";
+    undef $warn;
+    my $foo = POSIX::tmpnam();
+    is($warn, undef, "... but the warning can be disabled");
+}
+
 # Check that output is not flushed by _exit. This test should be last
 # in the file, and is not counted in the total number of tests.
 if ($^O eq 'vos') {
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @khwilliamson

On 07/21/2014 08&#8203;:02 PM\, Tony Cook via RT wrote&#8203;:
> On Sun Jun 01 06&#8203;:27&#8203;:39 2014\, jhi wrote&#8203;:
>> The tmpnam\(\) interface is known to be dangerous enough that compilers
>> and even linkers warn against its use\.  And POSIX\.pod advises
>> likewise\.
>>
>> Therefore\, POSIX extension probably should stop giving access to the
>> raw
>> tmpnam\(\)\.
>>
>> Some options&#8203;:
>>
>> \(1\) remove it\, let users \(\*\) die
>>   \(2\) implement it using mkstemp\(\) if available\.  \(What about platforms
>>   that have tmpnam\(\) but do not have mkstemp\(\)?\)  This reimplementing\,
>> of
>>   course\, is already done \(and much more\) in File&#8203;::Temp\, but we
>> probably
>> don't want the POSIX being dependent on File&#8203;::Temp\.
>>
>> Both options above can include "add a loud obnoxious warning"\.
>>
>> \(\*\) http&#8203;://grep\.cpan\.me/?q=POSIX\.\*tmpnam
>
> \(3\) produce a deprecation warning\, patch attached\.
>
> Tony

I have no objections to a deprecation warning\.  Note that there is code 
already in Posix\.xs to emit only one warning perl call site of functions 
that have it\.  Do we want a warning every time tmpnam is called\, or just 
the first time from any given place in the code?  If the latter\, the 
code already there could serve as a template\, or it could be extracted 
into a function and used from more than one place

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Jul 22 20&#8203;:40&#8203;:38 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 07/21/2014 08&#8203;:02 PM\, Tony Cook via RT wrote&#8203;:
> > On Sun Jun 01 06&#8203;:27&#8203;:39 2014\, jhi wrote&#8203;:
> >> The tmpnam\(\) interface is known to be dangerous enough that compilers
> >> and even linkers warn against its use\.  And POSIX\.pod advises
> >> likewise\.
> >>
> >> Therefore\, POSIX extension probably should stop giving access to the
> >> raw
> >> tmpnam\(\)\.
> >>
> >> Some options&#8203;:
> >>
> >> \(1\) remove it\, let users \(\*\) die
> >>   \(2\) implement it using mkstemp\(\) if available\.  \(What about platforms
> >>   that have tmpnam\(\) but do not have mkstemp\(\)?\)  This reimplementing\,
> >> of
> >>   course\, is already done \(and much more\) in File&#8203;::Temp\, but we
> >> probably
> >> don't want the POSIX being dependent on File&#8203;::Temp\.
> >>
> >> Both options above can include "add a loud obnoxious warning"\.
> >>
> >> \(\*\) http&#8203;://grep\.cpan\.me/?q=POSIX\.\*tmpnam
> >
> > \(3\) produce a deprecation warning\, patch attached\.
> >
> > Tony
> 
> I have no objections to a deprecation warning\.  Note that there is code 
> already in Posix\.xs to emit only one warning perl call site of functions 
> that have it\.  Do we want a warning every time tmpnam is called\, or just 
> the first time from any given place in the code?  If the latter\, the 
> code already there could serve as a template\, or it could be extracted 
> into a function and used from more than one place

That code isn't thread safe\, at least on Win32\.

The following code crashes in blead&#8203;:

use strict;
use threads;
use POSIX;

my $t = threads\->create\(sub \{ isupper\("a"\) \}\);
$t\->join;

islower\("a"\);

my $t2 = threads\->create\(sub \{ isupper\("a"\) \}\);

$t2\->join;

I believe it happens because the ptr\_table\_\*\(\) code uses Newx\(\) which uses safemalloc\(\)\, which on Win32\, allocates from a thread specific pool which is released when the thread exits\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

[is_common-backtrace.png](https://rt-archive.perl.org/perl5/Ticket/Attachment/1303521/691057/is_common-backtrace.png)

**Comment by p5pRT** [comment]
### From @tonycoz

On Tue Jul 22 20&#8203;:40&#8203;:38 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> I have no objections to a deprecation warning\.  Note that there is code 
> already in Posix\.xs to emit only one warning perl call site of functions 
> that have it\.  Do we want a warning every time tmpnam is called\, or just 
> the first time from any given place in the code?  If the latter\, the 
> code already there could serve as a template\, or it could be extracted 
> into a function and used from more than one place
> 

I've attached an updated patch that avoids the duplicate warnings\.

Tony

**Comment by p5pRT** [comment]
### From @tonycoz

<details><summary>0001-deprecate-POSIX-tmpnam.patch</summary>

```diff
From 567534033423e68f2d7a8e8810d3acc8c1ccaf83 Mon Sep 17 00:00:00 2001
From: Tony Cook <tony@develop-help.com>
Date: Thu, 2 Oct 2014 13:40:22 +1000
Subject: deprecate POSIX::tmpnam

This patch avoids repeating the deprecation warning if the same tmpnam()
call is executed multiple times.
---
 ext/POSIX/POSIX.xs     |    7 +++++++
 ext/POSIX/lib/POSIX.pm |    2 +-
 ext/POSIX/t/posix.t    |   17 ++++++++++++++++-
 3 files changed, 24 insertions(+), 2 deletions(-)

diff --git a/ext/POSIX/POSIX.xs b/ext/POSIX/POSIX.xs
index 1757549..5f546f5 100644
--- a/ext/POSIX/POSIX.xs
+++ b/ext/POSIX/POSIX.xs
@@ -2912,6 +2912,13 @@ tmpnam()
 	 *
 	 * Then again, maybe this should be removed at some point.
 	 * No point in enabling dangerous interfaces. */
+        if (ckWARN_d(WARN_DEPRECATED)) {
+	    HV *warned = get_hv("POSIX::_warned", GV_ADD | GV_ADDMULTI);
+            if (! hv_exists(warned, (const char *)&PL_op, sizeof(PL_op))) {
+                Perl_warner(aTHX_ packWARN(WARN_DEPRECATED), "Calling POSIX::tmpnam() is deprecated");
+                hv_store(warned, (const char *)&PL_op, sizeof(PL_op), &PL_sv_yes, 0);
+            }
+        }
 	len = strlen(tmpnam(SvPV(RETVAL, i)));
 	SvCUR_set(RETVAL, len);
     OUTPUT:
diff --git a/ext/POSIX/lib/POSIX.pm b/ext/POSIX/lib/POSIX.pm
index be1f097..4770a54 100644
--- a/ext/POSIX/lib/POSIX.pm
+++ b/ext/POSIX/lib/POSIX.pm
@@ -4,7 +4,7 @@ use warnings;
 
 our ($AUTOLOAD, %SIGRT);
 
-our $VERSION = '1.44';
+our $VERSION = '1.45';
 
 require XSLoader;
 
diff --git a/ext/POSIX/t/posix.t b/ext/POSIX/t/posix.t
index 398928c..18ea0be 100644
--- a/ext/POSIX/t/posix.t
+++ b/ext/POSIX/t/posix.t
@@ -8,7 +8,7 @@ BEGIN {
     }
 }
 
-use Test::More tests => 117;
+use Test::More tests => 120;
 
 use POSIX qw(fcntl_h signal_h limits_h _exit getcwd open read strftime write
 	     errno localeconv dup dup2 lseek access);
@@ -433,6 +433,21 @@ SKIP: {
     cmp_ok($!, '==', POSIX::ENOTDIR);
 }
 
+{   # tmpnam() is deprecated
+    my @warn;
+    local $SIG{__WARN__} = sub { push @warn, "@_"; note "@_"; };
+    my $x = sub { POSIX::tmpnam() };
+    my $foo = $x->();
+    $foo = $x->();
+    is(@warn, 1, "POSIX::tmpnam() should warn only once per location");
+    like($warn[0], qr!^Calling POSIX::tmpnam\(\) is deprecated at t/posix.t line \d+\.$!,
+       "check POSIX::tmpnam warns by default");
+    no warnings "deprecated";
+    undef $warn;
+    my $foo = POSIX::tmpnam();
+    is($warn, undef, "... but the warning can be disabled");
+}
+
 # Check that output is not flushed by _exit. This test should be last
 # in the file, and is not counted in the total number of tests.
 if ($^O eq 'vos') {
-- 
1.7.10.4


```
</details>

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed Oct 01 20&#8203;:46&#8203;:28 2014\, tonyc wrote&#8203;:
> On Tue Jul 22 20&#8203;:40&#8203;:38 2014\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > I have no objections to a deprecation warning\.  Note that there is code 
> > already in Posix\.xs to emit only one warning perl call site of functions 
> > that have it\.  Do we want a warning every time tmpnam is called\, or just 
> > the first time from any given place in the code?  If the latter\, the 
> > code already there could serve as a template\, or it could be extracted 
> > into a function and used from more than one place
> > 
> 
> I've attached an updated patch that avoids the duplicate warnings\.

Applied as cae71c5daa29ee5b24cc6ad145eadb323c1e1612\.

Tony

**Comment by p5pRT** [comment]
@tonycoz - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @rjbs

Jarkko suggests we remove this feature in 5\.24\, doing a one\-release deprecation\, because tmpnam is a known security problem\.

I can see both sides of this issue\.  I am about 0\.5% on the side of keeping it in for two releases\, as is standard practice\, to ensure that people see warnings at least once\.  The notion being that while perl is on a 1 year release cycle\, many downstream vendors only sync up one every other year\.

If we drop the code without two years of deprecation\, some upgraders will have their library fail to compile \(or run\) without warning ahead of time\.

Then again\, to what extent is the code really "working" now?  My sentiment is "maybe well enough\," but it's a lousy sentiment and I'm not really married to it\.  Anybody else want to throw an argument on the fire?

\-\- 
rjbs
